PROTON-1940: [c] normalize encoding of multiple="true" fields

alanconway opened the pull request:
AMQP spec allows several ways to encode a "multiple" field, we allow all in
incoming data or pn_data_t values created in the user, but when writing
to the wire we always encode:

- empty array as null
- single value array as a single value
- multiple values - encoded as array

This is the most compact encoding and seems to be best for interop, as some AMQP
clients (.NET) have trouble with an empty array in a multiple value field.

astitcher commented on c/src/core/codec.c:
```diff
@@ -657,7 +657,46 @@ int pn_data_vfill(pn_data_t *data, const char *fmt, va_list ap)
-    default:
+     case 'M': {                /* multiple=true field value */
+       /* Encode a pn_data_t * as a normalized multiple="true" field.
+          See file:///home/aconway/specs/amqp-core-complete-v1.0.pdf
```
What is 'multiple="true" field'? There is nothing called this in the standard is there? The ="true' naming seems to indicate boolean fields, but this is not the case afaict - and this really applies to a list of things that can also be empty in which case we just use null instead of the empty list.

Also url to local file system is not very helpful!

astitcher commented on c/src/core/codec.c:
```diff
@@ -529,15 +529,15 @@ int pn_data_vfill(pn_data_t *data, const char *fmt, va_list ap)
       err = pn_data_put_double(data, va_arg(ap, double));
       break;
-    case 'Z':
+    case 'Z':                   /* encode binary, must not be NULL */
```
As much as I agree the format letters need documenting I think that doing it inline is *not* the right way to do it - IMO a single block comment above would be much easier to encompass all the different codes.

alanconway commented:
Changes prompted by astitcher's comments:

- better explanation of mulitple fields, corrected AMQP spec links
- added block comment for vfill format letters
- factored out pni_normalize_multiple to reduce visual churn in _vfill() impl

Not changed: still using a new format letter. 2 reasons - consistency and efficiency/ease of impl:

consistency: the multiple attribute is part of an AMQP composite field definition. The purpose of _vfill() is to fill in composite fields correctly - so IMO that is where this belongs. I personally hate the vfill char-based mini-language; it is hard to read, hard to remember and error-prone. However it seems like something we should replace entirely or live with for now, rather than have inconsistent approaches for different fields.

ease of impl: as part of vfill, we can copy exactly as much of the source data as we need to the target. A normalize function outside of vfill would either have to modify the source value in-place (side-effects on caller's data value) or create an intermediate copy (expensive, extra memory management) Both are inconsistent with existing use of vfill.

alanconway commented:
Merged, if there are still objections to using a vfill code lets open a new issue
